### PR TITLE
Some cocoa refactoring

### PR DESCRIPTION
**Pull request by Akemi**
i will change the commit messages later to add line breaks, since the commits probably need some discussion/changes before merging anyway.

anyway i tried to understand the mechanics of the code and i am not quite sure if i could grasp everything properly or i just plainly forgot about a case. the original problem i stumbled upon, after my recent changes for the refresh rate retrieval, was that in the worst case mpv queried the display refresh rate up to 4 times at startup. which made everything unnecessary slow. furthermore it updates the refresh rate in cases it wasn't needed at all. some cases seem to have gotten obsolete by recent changes so i removed or changed them a bit.

I agree that my changes can be relicensed to LGPL 2.1 or later.

### CODE REVIEW

File: video/out/cocoa_common.m [hunk 1]
```diff
@@ -574,7 +574,6 @@ int vo_cocoa_config_window(struct vo *vo)
         if (!s->embedded && s->window) {
             if (reset_size)
                 queue_new_video_size(vo, width, height);
-            vo_cocoa_fullscreen(vo);
```
Comment (pigoz) [inline video/out/cocoa_common.m]: This is needed to handle `--fs`, IIRC. Does it still work without the line?

**Comment by pigoz** [comment]
About `windowDidChangeScreen`, keep in mind that cocoa creates a second borderless window when going FS, and moves the view of our window to that.

So I assume the delegate is not called when going FS with `--fs-screen != current`? (Same goes for `fs_screen_profile_observer`)

**Comment by Akemi** [comment]
i tested that case with --screen=0 --fs-screen=1. started mpv and then switched to fullscreen. the [commit](https://github.com/Akemi/mpv/commit/3a7d4971011d3e16e79a0dc8305bf4a13b99d7ac?diff=unified) for outputting this, just applied to the branch of this pull request. it actually triggers windowDidChangeScreen even two times.

```
<----- start ------>
[vo/opengl] vo_cocoa_control_get_icc_profile
[vo/opengl] vo_cocoa_config_window
[vo/opengl] UPDATE FPS 59.949785
<----- fullscreen ------>
[vo/opengl] VOCTRL_FULLSCREEN
[vo/opengl] vo_cocoa_fullscreen
[vo/opengl] UPDATE FPS 50.000000
[vo/opengl] windowDidChangeScreen
[vo/opengl] UPDATE FPS 50.000000
[vo/opengl] windowDidChangeScreen
[vo/opengl] vo_cocoa_control_get_icc_profile
```
the same just with master.
```
<----- start ------>
[vo/opengl] vo_cocoa_control_get_icc_profile
[vo/opengl] UPDATE FPS 59.949785
[vo/opengl] vo_cocoa_config_window
[vo/opengl] UPDATE FPS 59.949785
[vo/opengl] vo_cocoa_fullscreen
[vo/opengl] UPDATE FPS 59.949785
[vo/opengl] vo_cocoa_control_get_icc_profile
[vo/opengl] UPDATE FPS 59.949785

<----- fullscreen ------>
[vo/opengl] VOCTRL_FULLSCREEN
[vo/opengl] vo_cocoa_fullscreen
[vo/opengl] UPDATE FPS 50.000000
[vo/opengl] UPDATE FPS 50.000000
[vo/opengl] windowDidChangeScreen
[vo/opengl] UPDATE FPS 50.000000
[vo/opengl] windowDidChangeScreen
[vo/opengl] vo_cocoa_control_get_icc_profile
[vo/opengl] UPDATE FPS 50.000000

<----- windowed ------>
[vo/opengl] VOCTRL_FULLSCREEN
[vo/opengl] vo_cocoa_fullscreen
[vo/opengl] UPDATE FPS 59.949785
[vo/opengl] UPDATE FPS 59.949785
[vo/opengl] windowDidChangeScreen
[vo/opengl] vo_cocoa_control_get_icc_profile
[vo/opengl] UPDATE FPS 59.949785
```

do you want me to make a separate pull request for the last commit "cocoa: update screen fps only if really needed"? so it can be merged already since the rest of this still needs a bit of work. i also found another bug i introduced and should fix beforehand.

File: video/out/cocoa_common.m (hunk 1 above)
Comment (Akemi) [inline video/out/cocoa_common.m]: yes you are right about this, it doesn't work anymore. think it's worth to only call it when "vo->opts->fullscreen" is true? that way we can avoid unnecessary calls.

Comment (pigoz) [inline video/out/cocoa_common.m]: I am not sure, maybe @wm4 can tell more, but other backends call it unconditionally at start

Comment (wm4) [inline video/out/cocoa_common.m]: That function is completely private to cocoa, so I don't understand the question.

Just make sure ``--fs`` and switching FS at runtime with ``f`` works.

Comment (Akemi) [inline video/out/cocoa_common.m]: the actual problem imo is not the `vo_cocoa_fullscreen` call but some calls inside that function. specifically `vo_cocoa_update_screen_info()` and `flag_events(vo,VO_EVENT_ICC_PROFILE_CHANGED)`. former is already called at the beginning of `vo_cocoa_config_window` and just leads to redundancy, latter shouldn't be handled by the fullscreen code i would say (but i removed that in one of the commits). 

it's probably better to remove that commit and change `vo_cocoa_fullscreen` a bit?

**Comment by pigoz** [comment]
@Akemi sorry I forgot to reply. Yes please send a single PR for  "cocoa: update screen fps only if really needed" and I will merge that immediately.

**Comment by Akemi** [comment]
no worries i am a bit busy atm anyway. also opened a new PR.

though i am a bit unsure about what to do with the rest of the commits. i will revise `cocoa: remove one unneeded vo_cocoa_fullscreen call` since that commit break `--fs`.

for `cocoa: remove fs_screen_profile_observer ` i thought `cocoa_add_fs_screen_profile_observer` was for the case when current screen and fs screen are different but when i tested it a bit more i was completely unable to trigger that event/notification. i have the feeling that part is unfunctional and can be removed?

**Comment by wm4** [comment]
So is this PR outdated?

**Comment by Akemi** [comment]
yep i will update it since the other one was merged now.

**Comment by Akemi** [comment]
updated the PR. i removed the fs commit since it broke --fs.

**Comment by wm4** [comment]
Should this be merged? Does pigoz even care at all?

**Pull request closed without merging**